config/areas/users/dialogs.php

```diff
@@ -336,4 +336,69 @@
 		'submit'  => $fields['file']['submit']
 	],
 
+	// user disable TOTP
+	'user.totp.disable' => [
+		'pattern' => 'users/(:any)/totp/disable',
+		'load' => function (string $id) {
+			$user        = Find::user($id);
+			$currentUser = App::instance()->user();
+			$dialog      = [
+				'props' => [
+					'submitButton' => [
+						'text'  => I18n::translate('disable'),
+						'icon'  => 'protected',
+						'theme' => 'negative'
+					],
+				]
+			];
+
+			if (
+				$currentUser->isAdmin() === true &&
+				$currentUser->is($user) === false
+			) {
+				$dialog['component'] = 'k-remove-dialog';
+				$dialog['props']['text'] = I18n::template('login.totp.disable.admin', ['user' => $user->name()->or($user->email())]);
+			} else {
+				$dialog['component'] = 'k-form-dialog';
+				$dialog['props']['fields'] = [
+					'password' => [
+						'type'     => 'password',
+						'required' => true,
+						'counter'  => false,
+						'label'    => I18n::translate('login.totp.disable.label'),
+						'help'     => I18n::translate('login.totp.disable.help'),
+					]
+				];
+			}
+
+			return $dialog;
+		},
+		'submit' => function (string $id) {
+			$user     = Find::user($id);
+			$kirby    = App::instance();
+			$password = $kirby->request()->get('password');
+
+			try {
+				// Admins can disable TOTP without password
+				if (
+					$kirby->user()->isAdmin() === false ||
+					$kirby->user()->is($user) === true
+				) {
+					$user->validatePassword($password);
+				}
```

Comment:
I think it would be good to immediately block other non-admin users from calling this dialog submit route at all. With the current code there could be the case of another user (non-admin) who knows the password of another user but not the TOTP code – they could then call the route manually and provide the password.

I know there's another protection in `UserRules`, but another layer would IMO be good in such a critical part of the system.

So there should be three cases:

- Current user: Validate password (doesn't matter if admin or not)
- Admin: Don't validate password
- All other cases: Error

Reply:
Does this solve what you mean?

```php
if ($kirby->user()->is($user) === true) {
	$user->validatePassword($password);
} else if ($kirby->user()->isAdmin() === false) {
	throw new PermissionException('You are not allowed to disable TOTP for other users');
}
```

Reply:
Yes, that looks great!